conjure-go-client/httpclient/client_params.go

```diff
@@ -542,13 +542,38 @@ func WithErrorDecoder(errorDecoder ErrorDecoder) ClientParam {
 // WithBasicAuth sets the request's Authorization header to use HTTP Basic Authentication with the provided username and
 // password.
 func WithBasicAuth(user, password string) ClientOrHTTPClientParam {
-	return WithMiddleware(&basicAuthMiddleware{provider: func(ctx context.Context) (BasicAuth, error) {
-		return BasicAuth{User: user, Password: password}, nil
-	}})
+	return WithMiddleware(MiddlewareFunc(func(req *http.Request, next http.RoundTripper) (*http.Response, error) {
+		setBasicAuth(req.Header, user, password)
+		return next.RoundTrip(req)
+	}))
 }
 
+// WithBasicAuthProvider sets the request's Authorization header to use HTTP Basic Authentication.
+// The provider is expected to always return a nonempty BasicAuth value, or an error.
 func WithBasicAuthProvider(provider BasicAuthProvider) ClientOrHTTPClientParam {
-	return WithMiddleware(&basicAuthMiddleware{provider: provider})
+	return WithMiddleware(MiddlewareFunc(func(req *http.Request, next http.RoundTripper) (*http.Response, error) {
+		basicAuth, err := provider(req.Context())
+		if err != nil {
+			return nil, err
+		}
+		setBasicAuth(req.Header, basicAuth.User, basicAuth.Password)
+		return next.RoundTrip(req)
+	}))
```

Comment:
This is likely fine, but could also reduce some code duplication with:

```suggestion
        return WithBasicAuthOptionalProvider(BasicAuthOptionalProvider(func(context.Context) (*BasicAuth, error) {
        basicAuth, err := provider(ctx)
        if err != nil {
            return nil, err
        }
        return &basicAuth, nil
})
```